File: GridPatternExample/src/main/java/org/akka/essentials/grid/worker/WorkerActor.java
package org.akka.essentials.grid.worker;

import org.akka.essentials.grid.Task;
import org.akka.essentials.grid.TaskFinished;

import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;

public class WorkerActor extends UntypedActor {
	LoggingAdapter log = Logging.getLogger(getContext().system(), this);
	private static int instanceCounter = 0;
	private static int messageProcessedCounter = 0;

	@Override
	public void preStart() {
		instanceCounter++;
		log.info("Starting WorkerActor instance #" + instanceCounter
				+ ", hashcode #" + this.hashCode());
	}

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof String) {
			if (((String) message).compareTo("STOP") == 0) {
				// used when the new routee gets added to the server
				// we stop the currently running actor's on each client
				getContext().stop(getSelf());
			}
		} else if (message instanceof Task) {
			Task task = (Task) message;
			// check whether we have processed enough messages
			if (messageProcessedCounter == 100) {
				// tell the server, enough messages for me and shut me down
				getContext().stop(getSelf());
				getContext().system().shutdown();
			} else {
				log.info("Work Packet from server->" + task.getTaskNumber());
				messageProcessedCounter++;
				getSender().tell(new TaskFinished(task.getTaskNumber()));
			}
		}
	}

	@Override
	public void postStop() {
		log.info("Stopping WorkerActor instance #" + instanceCounter
				+ ", hashcode #" + this.hashCode());
		instanceCounter--;
	}

}

File: GridPatternExample/src/main/scala/org/akka/essentials/grid/worker/WorkerActor.scala
package org.akka.essentials.grid.worker
import org.akka.essentials.grid.controller.ENOUGH
import org.akka.essentials.grid.controller.STOP
import org.akka.essentials.grid.controller.Task
import org.akka.essentials.grid.controller.TaskFinished

import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorLogging

/*
 * the message processed counter gets reset when the grid controller starts and stops 
 * the worker instance #bug
 * 
 */
class WorkerActor extends Actor with ActorLogging {
  var instanceCounter: Int = 0
  var messageProcessedCounter: Int = 0

  override def preStart() {
    instanceCounter += 1
    log.info("Starting WorkerActor instance #" + instanceCounter
      + ", hashcode #" + this.hashCode())
  }
  override def postStop() {
    log.info("Stopping WorkerActor instance #" + instanceCounter
      + ", hashcode #" + this.hashCode())
    instanceCounter -= 1
  }
  def receive: Receive = {
    case STOP =>
      // used when the new routee gets added to the server
      // we stop the currently running actor's on each client
      context.stop(self)

    case task: Task =>
      // check whether we have processed enough messages
      if (messageProcessedCounter == 10) {
        // tell the server, enough messages for me and shut me down
        sender ! new ENOUGH
        context.stop(self)
        context.system.shutdown()
      } else {
        log.info("Work Packet from server->" + task.taskNumber)
        messageProcessedCounter += 1;
        sender ! new TaskFinished(task.taskNumber)
      }
  }
}